app/models/blocked_object.rb

```diff
@@ -14,16 +14,17 @@ class BlockedObject
   field :expires_at, type: DateTime
   field :blocked_by, type: Integer
 
-  scope :email,                        -> { where(object_type: BLOCKED_OBJECT_TYPES[:email]) }
-  scope :email_domain,                 -> { where(object_type: BLOCKED_OBJECT_TYPES[:email_domain]) }
-  scope :browser_guid,                 -> { where(object_type: BLOCKED_OBJECT_TYPES[:browser_guid]) }
-  scope :ip_address,                   -> { where(object_type: BLOCKED_OBJECT_TYPES[:ip_address]) }
-  scope :charge_processor_fingerprint, -> { where(object_type: BLOCKED_OBJECT_TYPES[:charge_processor_fingerprint]) }
-  scope :product,                      -> { where(object_type: BLOCKED_OBJECT_TYPES[:product]) }
-  scope :active,                       -> { where(:blocked_at.ne => nil).any_of({ expires_at: nil }, { :expires_at.gt => Time.current }) }
-
+  BLOCKED_OBJECT_TYPES.each_value do |object_type|
+    scope object_type, -> { where(object_type:) }
+    define_method("#{object_type}?") { self.object_type == object_type }
+  end
   validates_inclusion_of :object_type, in: BLOCKED_OBJECT_TYPES.values
-  validate :expires_on_required_for_active_object_type_ip_address
+  validates :expires_at, presence: {
+    if: [:ip_address?, :blocked_at?],
+    unless: :expires_at?
```

Comment:
we can remove this `unless` because it's already checked using `presence` above? or do we want to skip validating `expires_at` if it's already set?

Reply:
Makes sense!